### DISCUSSION TITLE
Regular Expression manager for updating trivy

### DISCUSSION THREAD

**Discussion by MShekow** [renovatebot/renovate#15413] [Request Help] [answered]
Hi,

how can I write a `regexManager` that allows me to update the Trivy dependency that I download e.g. with `curl` or `wget` from [here](https://github.com/aquasecurity/trivy/releases) (such as `https://github.com/aquasecurity/trivy/releases/download/v0.27.1/trivy_0.27.1_Linux-64bit.tar.gz` )? You will quickly notice that an answer such as the one below is utterly useless ;)

```
{
  "fileMatch": ["^.gitlab-ci.yml$"],
  "matchStrings": ["wget https:\/\/github\.com\/(?<depName>[a-z]+\/[a-z]+)\/releases\/download\/(?<currentValue>.*?)\/.* -O"],
  "datasourceTemplate": "github-releases"
}
```

The reason why this doesn't work is obvious: Trivy declares the dependencies by versioning it with the `v` prefix, but then also puts the version (**without** the `v`) into the URL a second time (into the file name).

The following also doesn't work, because I get an error (`Invalid regExp for regexManagers`), presumably because `(?<currentValue>.*?)` is part of the regEx _twice_.

```
{
      "fileMatch": ["^.gitlab-ci.yml$"],
      "matchStrings": [
        "wget https:\/\/github\.com\/aquasecurity\/trivy\/releases\/download\/v(?<currentValue>.*?)\/trivy_(?<currentValue>.*?)_Linux-64bit\.tar\.gz -O"
      ],
      "datasourceTemplate": "github-releases",
      "versioningTemplate": "regex:^v(?<major>\\d+)(\\.(?<minor>\\d+))(\\.(?<patch>\\d+))$",
      "depNameTemplate": "aquasecurity/trivy"
    }
```

Any suggestions?

**Comment by viceice** [comment, marked as answer]
use a environment variable, sample

https://github.com/visualon/docker-images/blob/main/windows/2019/drone-cache/Dockerfile#L3

**Reply by MShekow** [reply]
Thanks for the pointer. I turned it into a minimal example that works for GitLab CI/CD pipelines:

`.gitlab-ci.yml`:
```
variables:
    # renovate: datasource=github-releases depName=aquasecurity/trivy
    TRIVY_VERSION: v0.27.0
some job:
    stage: build
    image: alpine:latest
    script:
        - apk add -u wget
        - 'wget --no-verbose https://github.com/aquasecurity/trivy/releases/download/${TRIVY_VERSION}/trivy_${TRIVY_VERSION#v}_Linux-64bit.tar.gz -O - | tar -zxvf -'
```
Note the `#v` to get rid of the the `v` prefix. ;)

`renovate.json5`:
```
"regexManagers": [
    {
      "fileMatch": ["^.gitlab-ci.yml$"],
      "matchStrings": [
        "renovate: datasource=(?<datasource>.*?) depName=(?<depName>.*?)( versioning=(?<versioning>.*?))?\\s*.*?_VERSION: (?<currentValue>.*)\\s"
      ],
      "versioningTemplate": "{{#if versioning}}{{{versioning}}}{{else}}semver{{/if}}"
    }
  ]

```
